crates/matrix-sdk/tests/integration/widget.rs

```diff
@@ -0,0 +1,228 @@
+use tracing::error;
+use wiremock::{
+    matchers::{header, method, path_regex, query_param},
+    Mock, MockServer, ResponseTemplate,
+};
+
+use crate::{logged_in_client, mock_sync};
+
+/// Create a JSON string from a [`json!`][serde_json::json] "literal".
+#[macro_export]
+macro_rules! json_string {
+    ($( $tt:tt )*) => { ::serde_json::json!( $($tt)* ).to_string() };
+}
+
+const WIDGET_ID: &str = "test-widget";
+static ROOM_ID: Lazy<OwnedRoomId> = Lazy::new(|| owned_room_id!("!a98sd12bjh:example.org"));
+
+async fn run_test_driver() -> (MockServer, WidgetDriverHandle) {
+    struct DummyPermissionsProvider;
+
+    #[async_trait]
+    impl PermissionsProvider for DummyPermissionsProvider {
+        async fn acquire_permissions(&self, permissions: Permissions) -> Permissions {
+            // Grant all permissions that the widget asks for
+            permissions
+        }
+    }
+
+    let (client, mock_server) = logged_in_client().await;
+    let sync_settings = SyncSettings::new().timeout(Duration::from_millis(3000));
+
+    let mut ev_builder = SyncResponseBuilder::new();
+    ev_builder.add_joined_room(JoinedRoomBuilder::new(ROOM_ID.clone()));
+
+    mock_sync(&mock_server, ev_builder.build_json_sync_response(), None).await;
+    let _response = client.sync_once(sync_settings.clone()).await.unwrap();
+    mock_server.reset().await;
+
+    let room = client.get_room(&ROOM_ID).unwrap();
+
+    let (driver, handle) = WidgetDriver::new(
+        WidgetSettings::new(WIDGET_ID.to_owned(), true, "https://foo.bar/widget").unwrap(),
+    );
+
+    spawn(async move {
+        if let Err(()) = driver.run(room, DummyPermissionsProvider).await {
+            error!("An error encountered in running the WidgetDriver (no details available yet)");
+        }
+    });
+
+    (mock_server, handle)
+}
+
+async fn recv_message(driver_handle: &WidgetDriverHandle) -> JsonObject {
+    serde_json::from_str(&driver_handle.recv().await.unwrap()).unwrap()
+}
+
+async fn send_request(
+    driver_handle: &WidgetDriverHandle,
+    request_id: &str,
+    action: &str,
+    data: impl Serialize,
+) {
+    let sent = driver_handle
+        .send(json_string!({
+            "api": "fromWidget",
+            "widgetId": WIDGET_ID,
+            "requestId": request_id,
+            "action": action,
+            "data": data,
+        }))
+        .await;
+    assert!(sent);
+}
+
+async fn send_response(
+    driver_handle: &WidgetDriverHandle,
+    request_id: &str,
+    action: &str,
+    request_data: impl Serialize,
+    response_data: impl Serialize,
+) {
+    let sent = driver_handle
+        .send(json_string!({
+            "api": "toWidget",
+            "widgetId": WIDGET_ID,
+            "requestId": request_id,
+            "action": action,
+            "data": request_data,
+            "response": response_data,
+        }))
+        .await;
+    assert!(sent);
+}
+
+#[async_test]
+#[ignore = "widget API is not yet fully implemented"]
```

Comment:
If we're ignoring this test because it's failing some assertions: can we invert these, so the assertions show the current behavior, and will fail as soon as we fix the underlying issues, acting as a reminder to update the test? (and we don't end up forgetting that we `ignore`d the test)

Reply:
It's failing in a helper function `.unwrap()`ping things right now, so I'd rather not. I'll have to write more tests here soon enough so I don't think I'll forget.